### PR TITLE
Add a pass to tile and vectorize workgroup tiles.

### PR DESCRIPTION
**Pull request by asaadaldien**
- Tile and distribute linalg.batch_matmul.
- Tile and vectorize linalg.matmul and linalg.batch_matmul.
- Remove MatMulVectorization,
- Enable Distribute + Tile + Vectorize passes by default.
- Add llvm flag driven kernel dispatch with default parameters getting
lowest latency on mobile_bert benchmarks:

```
------------------------------------------------------------------------------------
Benchmark                                          Time             CPU
Iterations
------------------------------------------------------------------------------------
BM_serving_default/process_time/real_time        923 ms          918 ms
```

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+template <typename LinalgOpTy>
+struct TileWorkgroups : public linalg::LinalgBaseTilingPattern {
+  using Base = linalg::LinalgBaseTilingPattern;
+  TileWorkgroups(MLIRContext *context, linalg::LinalgTilingOptions options,
+                 linalg::LinalgMarker marker, PatternBenefit benefit = 1)
+      : Base(LinalgOpTy::getOperationName(), context, options, marker,
+             benefit) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    // Find the parent FuncOp before tiling. If tiling succeeds, the op will be
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp:39]: Comment seems off. The `LinalgBaseTilingPattern` doesnt erase the op (and you are erasing it here as you should). Also doesnt seem like you need `funcOp` at all.

File: iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+          [&cpuKernelDispatch](OpBuilder &builder,
+                               Operation *operation) -> SmallVector<Value, 4> {
+            return TileSizeFn::get<TilingLevel::Level2Tiles>(
+                cpuKernelDispatch, builder, operation);
+          }),
+      linalg::LinalgMarker(Identifier::get(getWorkgroupL1TileMarker(), context),
+                           Identifier::get(getVectorizeMarker(), context)));
+
+  // Apply tiling.
+  applyPatternsAndFoldGreedily(funcOp, std::move(l1patterns));
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp:102]: Not sure you need the `move`. The function expects a const &.

File: iree/compiler/Conversion/LinalgToLLVM/test/matmul_vectorization.mlir [hunk 3]
```diff
@@ -1,31 +1,35 @@
-// RUN: iree-opt --iree-codegen-linalg-to-llvm-matmul-vectorization-pass -split-input-file %s | IreeFileCheck %s
+// RUN: iree-opt -iree-codegen-llvm-linalg-tile-and-distribute -iree-codegen-linalg-to-llvm-workgroups-vectorization-pass -split-input-file %s | IreeFileCheck %s
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToLLVM/test/matmul_vectorization.mlir:1]: Nit: One of the reasons I dont like splitting the passes. Too many flags!

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp (hunk 1 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp]: Done!

File: iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp (hunk 2 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp:97]: This is needed, The API is update recently using ti use `FrozenRewritePatternList`
https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h#L38

Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp:97]: Ok, I missed that. My issue with such code style is that if you unknowingly use `l1Patterns` later you will get weird behavior.
Two options
1) Use FrozenRewritePatternList here and avoid using move.
2)  My preference is doing something like this
```
{
   OwningRewritePatternList l1PatternList;
   ...
   applyPatternsAndFoldGreedily(..., std::move(l1PatternList))
}
```
This way the list is out of scope when you move it.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp (hunk 2 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp]: Done!

File: iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        linalg::LinalgTilingOptions().setTileSizeComputationFunction(
+            [&cpuKernelDispatch](OpBuilder &builder, Operation *operation)
+                -> SmallVector<Value, 4> {
+              return TileSizeFn::get<TilingLevel::Level1Tiles>(
+                  cpuKernelDispatch, builder, operation);
+            }),
+        linalg::LinalgMarker(
+            Identifier::get(getWorkgroupMarker(), context),
+            Identifier::get(getWorkgroupL1TileMarker(), context)));
+
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp:83]: Nit: based on previous comment. Split the scope here. l2patterns not used before here.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**